ocaml/xapi/xapi_vm_clone.ml

```diff
@@ -498,6 +503,10 @@ let clone ?snapshot_info_record ?(ignore_vdis = []) disk_op ~__context ~vm
           let (_ : [`VGPU] Ref.t list) =
             List.map (fun vgpu -> Xapi_vgpu.copy ~__context ~vm:ref vgpu) vgpus
           in
+          (* copy vTPMs *)
+          let (_ : [`VTPM] Ref.t list) =
+            List.map (fun vtpm -> Xapi_vtpm.copy ~__context ~vM:ref vtpm) vtpms
```

Comment:
This looks strange. We are using `List.map` but are not using the result? Why not use `List.iter`? Or maybe use `ignore` and be more explicit?

Reply:
I also thought of using List.iter, but too keep it in 3 lines it needs the use of ignore, I'd rather use map with the explicit types.

Using iter and ignore would look like this, which is not too far off, but if copy changes we might not detect that change because we don't enforce types:
```
          List.iter
            (fun vtpm -> ignore (Xapi_vtpm.copy ~__context ~vM:ref vtpm))
            vtpms ;
```

Reply:
I believe `ignore` would detect the case where the function application becomes partial but I agree that types are not as explicit.

Reply:
Seems like all other object copies are implemented weirdly like this, I would've expected that you receive the original vtpm object, you make a copy and you put the ref that got returned into the new object.
Otherwise where does it go?

Reply:
Can we change the type of `copy` to return `unit`, and in the mli mention that it performs the copy and puts the new object into the specified target VM?
I think it is weird like this because it needs to maintain the relations in the DB: when you copy the VTPM object you also need to update its VM field which is a backreference to its owning VM (which you also changed when copying).

Reply:
It is useful to have all the `copy` methods return the reference of the newly created method in case the caller needs to do some operation with them (like the vtpm code does when it copies secrets). In this particular case that is not needed and the new references are simply ignored. This patterns is also followed across xapi for `create`.

> I think it is weird like this because it needs to maintain the relations in the DB: when you copy the VTPM object you also need to update its VM field which is a backreference to its owning VM (which you also changed when copying)
The VM is created by `copy_vm_record` which avoids giving any parameters of relationships to the actual `VM.create`.

I'd rather follow the current pattern rather than make the copy function return `unit` to make this callsite more pleasant to interpret.